File: backend/routes/reports.py
from __future__ import annotations

from flask import Blueprint, jsonify, request, current_app

from backend.services.report_service import ReportService

reports_bp = Blueprint("reports", __name__, url_prefix="/api/reports")


@reports_bp.get("/student/<int:student_id>")
def get_student_report(student_id: int):
    service = current_app.config.get("REPORT_SERVICE", ReportService)
    report = service.get_student_report(student_id)
    if not report:
        return jsonify({"error": "Student not found"}), 404
    return jsonify(report), 200


@reports_bp.get("/topic/<string:topic_id>")
def get_topic_report(topic_id: str):
    service = current_app.config.get("REPORT_SERVICE", ReportService)
    report = service.get_topic_report(topic_id)
    if not report:
        return jsonify({"error": "Topic not found"}), 404
    return jsonify(report), 200


@reports_bp.get("/class/overview")
def get_class_overview():
    service = current_app.config.get("REPORT_SERVICE", ReportService)
    overview = service.get_class_overview()
    return jsonify(overview), 200


@reports_bp.get("/question/<string:topic_id>/analytics")
def get_question_analytics(topic_id: str):
    subtopic_type = request.args.get("subtopic_type")
    service = current_app.config.get("REPORT_SERVICE", ReportService)
    analytics = service.get_question_analytics(topic_id, subtopic_type=subtopic_type)
    return jsonify(analytics), 200

File: src/services/reports.ts
import api from './api';

export interface StudentReport {
  student_id: number;
  student_name: string;
  student_email: string;
  overall_stats: {
    total_questions_answered: number;
    total_correct: number;
    total_incorrect: number;
    total_skipped: number;
    overall_accuracy: number;
    avg_time_per_question: number;
  };
  topic_breakdown: Array<{
    topic: string;
    topic_name: string;
    questions_answered: number;
    correct: number;
    incorrect: number;
    accuracy: number;
    avg_time: number;
    completion_percentage: number;
  }>;
  struggling_subtopics?: Array<{
    topic: string;
    subtopic_type: string;
    attempts: number;
    accuracy: number;
  }>;
}

export interface ClassOverview {
  total_students: number;
  active_students_last_week: number;
  total_questions_answered: number;
  class_avg_accuracy: number;
  topics_overview: Array<{
    topic: string;
    topic_name: string;
    students_started: number;
    students_completed: number;
    avg_accuracy: number;
  }>;
  rostered_students?: Array<{
    student_id: number | null;
    student_name: string;
    student_email: string;
  }>;
  top_performers: Array<{
    student_id: number;
    student_name: string;
    questions_answered: number;
    accuracy: number;
  }>;
  struggling_students: Array<{
    student_id: number;
    student_name: string;
    questions_answered: number;
    accuracy: number;
  }>;
  recent_activity?: Array<{
    date: string;
    questions_answered: number;
    active_students: number;
  }>;
}

export interface TopicReport {
  topic: string;
  topic_name: string;
  overall_stats: {
    total_students: number;
    students_started: number;
    students_completed: number;
    total_attempts: number;
    avg_accuracy: number;
    avg_time_per_question: number;
  };
  subtopic_difficulty: Array<{
    subtopic_type: string;
    attempts: number;
    unique_students: number;
    success_rate: number;
    avg_time: number;
    difficulty_rating: string;
  }>;
  most_missed_questions: Array<{
    question_code: string;
    subtopic_type: string;
    attempts: number;
    success_rate: number;
  }>;
}

export interface QuestionAnalyticsResponse {
  topic: string;
  analytics: Array<{
    question_code: string;
    subtopic_type: string;
    times_shown: number;
    correct_count: number;
    incorrect_count: number;
    skipped_count: number;
    success_rate: number;
    avg_time_spent: number;
    students_who_saw: number;
  }>;
}

export const reportsService = {
  async getStudentReport(studentId: number): Promise<StudentReport> {
    const response = await api.get(`reports/student/${studentId}`);
    return response.data;
  },

  async getClassOverview(): Promise<ClassOverview> {
    const response = await api.get('reports/class/overview');
    return response.data;
  },

  async getTopicReport(topicId: string): Promise<TopicReport> {
    const response = await api.get(`reports/topic/${topicId}`);
    return response.data;
  },

  async getQuestionAnalytics(topicId: string, subtopicType?: string): Promise<QuestionAnalyticsResponse> {
    const params = subtopicType ? { subtopic_type: subtopicType } : undefined;
    const response = await api.get(`reports/question/${topicId}/analytics`, { params });
    return response.data;
  },
};
